### PR TITLE
[CDF-22784] Postgres gateway tables

### PR DESCRIPTION
**Pull request by doctrino**
## Description
For disussion:

1. The dataclass `Column` is deviating from the API spec to increase usability.
2. Replace the `RawTableOptions` and `ViewTableOptions` with `RawTableId` and `ViewId`? Disadvantage: coupling, Advantage: DRY and maybe some useability.

Issue with docs: 
* Name collision TableList in PostgresGateway and TableList in Raw.

## Checklist:
- [ ] Tests added/updated.
- [ ] Documentation updated. Documentation is generated from docstrings - these must be updated according to your change.
  If a new method has been added it should be referenced in [cognite.rst](https://github.com/cognitedata/cognite-sdk-python/blob/master/docs/source/cognite.rst) in order to generate docs based on its docstring.
- [ ] Changelog updated in [CHANGELOG.md](https://github.com/cognitedata/cognite-sdk-python/blob/master/CHANGELOG.md).
- [ ] Version bumped. If triggering a new release is desired, bump the version number in [_version.py](https://github.com/cognitedata/cognite-sdk-python/blob/master/cognite/client/_version.py) and [pyproject.toml](https://github.com/cognitedata/cognite-sdk-python/blob/master/pyproject.toml) per [semantic versioning](https://semver.org/).

### CODE REVIEW

File: cognite/client/_api/postgres_gateway/tables.py [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+                >>> client = CogniteClient()
+                >>> table = ViewTableWrite(tablename="myCustom", options=ViewTableOptions(space="mySpace", external_id="myExternalId", version="v1"))
+                >>> res = client.postgres_gateway.tables.create(table, "myUserName")
+
+        """
+        self._warning.warn()
+        return self._create_multiple(
+            list_cls=TableList,
+            resource_cls=Table,  # type: ignore[type-abstract]
+            resource_path=self._RESOURCE_PATH.format(username=quote(username)),
```
Comment (haakonvt) [inline cognite/client/_api/postgres_gateway/tables.py:112]: Consider using `interpolate_and_url_encode` instead

File: cognite/client/data_classes/postgres_gateway/tables.py [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+        super().__init__(tablename=tablename)
+        self.options = options
+        self.columns = columns
+
+    @classmethod
+    def _load_table(cls, data: dict[str, Any]) -> Self:
+        return cls(
+            tablename=data["tablename"],
+            options=RawTableOptions._load(data["options"]),
+            columns=Column._load_columns(data["columns"]),
```
Comment (haakonvt) [inline cognite/client/data_classes/postgres_gateway/tables.py:167]: Would it make sense if this was a `ColumnList`?
```py
  columns=ColumnList._load(data["columns"]),
```
What I'm thinking is we could maybe drop the non-standard load and dump methods (`_dump_columns` / `_load_columns`)

**Review by haakonvt — COMMENTED** [review]
> 1. The dataclass Column is deviating from the API spec to increase usability.

No objections from me 👍 

> 2. Replace the RawTableOptions and ViewTableOptions with RawTableId and ViewId? Disadvantage: coupling, Advantage: DRY and maybe some useability.

Have you asked why they named it "options"? Seems more like what we would call id/identifier, so I agree with you there. I think maybe we should keep them separate, but no strong feelings.

**Comment by haakonvt** [comment]
> Name collision TableList in PostgresGateway and TableList in Raw.

I think this is fine; we have not added data classes for several new APIs into `cognite.client.data_classes`, and this is just one more. If this means we need to be more explicit when referring in the docs to make shinx happy that seems ok (good) to me 👍

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -53,7 +53,7 @@ repos:
       - id: pydoclint
         require_serial: true  # Spammy in run-all scenarios (more than fast enough already)
         args:
-          - --style=google
+          - --style=sphinx
```
Comment (doctrino) [inline .pre-commit-config.yaml:56]: Needed with the explicit reference to classes used in ProstgresGateway

File: cognite/client/data_classes/postgres_gateway/tables.py [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+    "TIMESTAMPTZ",
+    "JSON",
+    "TEXT[]",
+    "VARCHAR[]",
+    "BOOL[]",
+    "BIGINT[]",
+    "DOUBLE PRECISION[]",
+    "REAL[]",
+    "JSON[]",
+    "BYTEA",
```
Comment (haakonvt) [inline cognite/client/data_classes/postgres_gateway/tables.py:39]:
```suggestion
    "BIGINT",
    "BIGINT[]",
    "BOOL",
    "BOOL[]",
    "BYTEA",
    "DOUBLE PRECISION",
    "DOUBLE PRECISION[]",
    "JSON",
    "JSON[]",
    "REAL",
    "REAL[]",
    "TEXT",
    "TEXT[]",
    "TIMESTAMPTZ",
    "VARCHAR",
    "VARCHAR[]",
```

File: cognite/client/data_classes/postgres_gateway/tables.py [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+        if not isinstance(data, dict):
+            return cls._load(data, None)
+        columns = cls([], None)
+        for name, column_data in data.items():
+            columns.append(
+                Column(
+                    name=name,
+                    type=column_data["type"],
+                )
+            )
```
Comment (haakonvt) [inline cognite/client/data_classes/postgres_gateway/tables.py:91]: 👀

File: cognite/client/data_classes/postgres_gateway/tables.py [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+
+    Args:
+        tablename (str): Name of the foreign table.
+        options (RawTableOptions): Table options
+        columns (ColumnList): Foreign table columns.
+        created_time (int | None): Time when the table was created.
+
+    """
+
+    _type = "raw_rows"
```
Comment (haakonvt) [inline cognite/client/data_classes/postgres_gateway/tables.py:244]: This is also used by `RawTableWrite` - intended?

File: cognite/client/utils/_identifier.py [hunk 7]
```diff
@@ -384,6 +398,38 @@ def assert_singleton(self) -> None:
         if not self.is_singleton():
             raise ValueError("Exactly one username (string) must be specified")
 
+    @staticmethod
+    def unwrap_identifier(identifier: str | int | dict) -> str | int | InstanceId:
+        if isinstance(identifier, (str, int)):
+            return identifier
```
Comment (haakonvt) [inline cognite/client/utils/_identifier.py:404]: Can `Username` be int? 🤔 
```suggestion
        if isinstance(identifier, str):
            return identifier
```

File: cognite/client/utils/_identifier.py [hunk 8]
```diff
@@ -384,6 +398,38 @@ def assert_singleton(self) -> None:
...
+        raise TypeError(f"tablenames must be of type str or SequenceNotStr[str]. Found {type(tablenames)}")
+
+    def assert_singleton(self) -> None:
+        if not self.is_singleton():
+            raise ValueError("Exactly one tablename (string) must be specified")
+
+    @staticmethod
+    def unwrap_identifier(identifier: str | int | dict) -> str | int | InstanceId:
+        if isinstance(identifier, (str, int)):
+            return identifier
```
Comment (haakonvt) [inline cognite/client/utils/_identifier.py:428]:
```suggestion
        if isinstance(identifier, str):
            return identifier
```

File: tests/tests_integration/test_api/test_postgres_gateway/test_tables.py [hunk 9]
```diff
@@ -0,0 +1,128 @@
+import pytest
+from jaraco.context import suppress
```
Comment (haakonvt) [inline tests/tests_integration/test_api/test_postgres_gateway/test_tables.py:2]: Hmm, auto-imported by your IDE?
```suggestion
from contextlib import suppress
```

File: tests/tests_integration/test_api/test_postgres_gateway/test_tables.py [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+def one_raw_table(cognite_client: CogniteClient) -> tuple[str, str]:
+    db_table_pair = f"my_database-{random_string(10)}", "my_table_postgres_gateway"
+
+    cognite_client.raw.databases.create(db_table_pair[0])
+    cognite_client.raw.tables.create(*db_table_pair)
+
+    yield db_table_pair
+    with suppress(CogniteAPIError):
+        cognite_client.raw.tables.delete(*db_table_pair)
+        cognite_client.raw.databases.delete(db_table_pair[0])
```
Comment (haakonvt) [inline tests/tests_integration/test_api/test_postgres_gateway/test_tables.py:73]: If the first line raises `CogniteAPIError`, the second line will never be run. Is that intended?

File: tests/tests_integration/test_api/test_postgres_gateway/test_tables.py [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+            with pytest.raises(CogniteAPIError):
+                cognite_client.postgres_gateway.tables.retrieve(tablename, username)
+
+            result = cognite_client.postgres_gateway.tables.retrieve(tablename, username, ignore_unknown_ids=True)
+
+            assert result is None
+        finally:
+            if created:
+                with suppress(CogniteAPIError):
+                    # Bug in API, ignore_unknown_ids=True returns 500
```
Comment (haakonvt) [inline tests/tests_integration/test_api/test_postgres_gateway/test_tables.py:121]: Use xfail?

**Review by haakonvt — COMMENTED** [review]
Mostly LGTM! Does the POST_REGEX thingy need an update?

File: cognite/client/data_classes/postgres_gateway/tables.py (hunk 5 above)
Comment (doctrino) [reply cognite/client/data_classes/postgres_gateway/tables.py:91]: ?

File: cognite/client/data_classes/postgres_gateway/tables.py (hunk 6 above)
Comment (doctrino) [reply cognite/client/data_classes/postgres_gateway/tables.py:244]: I don't think I follow here. RawTableWrite also has type set

File: cognite/client/data_classes/postgres_gateway/tables.py (hunk 5 above)
Comment (haakonvt) [reply cognite/client/data_classes/postgres_gateway/tables.py:91]: 🏌️ ( 😆  )

File: cognite/client/data_classes/postgres_gateway/tables.py (hunk 6 above)
Comment (haakonvt) [reply cognite/client/data_classes/postgres_gateway/tables.py:244]: Just that they used the same value of `_type`, but I guess it's okay as they are subclasses of different classes

**Review by haakonvt — APPROVED** [review]
LGTM

**Pull request merged**